File: plugins/com.googlecode.contraildb.tests/src/com/googlecode/contraildb/tests/ContrailTasksTests.java
/*******************************************************************************
 * Copyright (c) 2009 Ted Stockwell
 * 
 * This file is part of the Contrail Database System.
 * 
 * Contrail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License Version 3
 * as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.googlecode.contraildb.tests;

import java.io.File;

import junit.framework.TestCase;

import com.google.appengine.tools.development.ApiProxyLocalFactory;
import com.google.apphosting.api.ApiProxy;
import com.googlecode.contraildb.core.ContrailServiceFactory;
import com.googlecode.contraildb.core.IContrailService;
import com.googlecode.contraildb.core.IContrailService.Mode;
import com.googlecode.contraildb.core.IContrailSession;
import com.googlecode.contraildb.core.Item;
import com.googlecode.contraildb.core.storage.provider.IStorageProvider;
import com.googlecode.contraildb.core.storage.provider.RamStorageProvider;


/**
 * task concurrency tests.
 * 
 * @author Ted Stockwell
 */
public class ContrailTasksTests extends TestCase {

	private IStorageProvider _storageProvider;
	private IContrailService _datastore;
	
	@Override
	protected void tearDown() throws Exception {
		_datastore.close();
	}
	
	@Override
	public void setUp() throws Exception {
		super.setUp();
		
		// setup Google App Engine environment
		ApiProxy.setEnvironmentForCurrentThread(new TestEnvironment());
		ApiProxyLocalFactory factory= new ApiProxyLocalFactory();
		factory.setApplicationDirectory(new File("."));
		ApiProxy.setDelegate(factory.create());

		
		_storageProvider= new RamStorageProvider();
		_datastore = ContrailServiceFactory.getContrailService(_storageProvider);
		
	}
	
	
	public void testDeadlock() throws Exception {
		IContrailSession session= _datastore.beginSession(Mode.READWRITE);
		Item object_0_1= new Item("person-0.1");
		session.store(object_0_1);
		session.commit();
		session.close();
		
		IContrailSession T1= _datastore.beginSession(Mode.READWRITE);
		
		// put object-1.1 in transaction 1
		Item object_1_1= new Item("person-1.1");
		T1.store(object_1_1);
		T1.delete(object_0_1.getId());
		
		assertTrue(T1.isActive());
		T1.close(); // abandon any changes
		assertFalse(T1.isActive());

		// object-1.1 should not be visible 
		// object-0.1 should still be visible 
		IContrailSession T3= _datastore.beginSession(Mode.READONLY);
		assertNull("transaction isolation violated", T3.fetch(object_1_1.getId()));
		assertNotNull("transaction isolation violated", T3.fetch(object_0_1.getId()));
		T3.close();
		IContrailSession T4= _datastore.beginSession(Mode.READWRITE);
		assertNull("transaction isolation violated", T4.fetch(object_1_1.getId()));
		assertNotNull("transaction isolation violated", T4.fetch(object_0_1.getId()));
		T4.close();
	}
	
	
}

File: branches/scala/plugins/com.googlecode.contraildb.tests/src/com/googlecode/contraildb/tests/ContrailTasksTests.scala
/*******************************************************************************
 * Copyright (c) 2009 Ted Stockwell
 * 
 * This file is part of the Contrail Database System.
 * 
 * Contrail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License Version 3
 * as published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.googlecode.contraildb.tests;

import java.util.concurrent.atomic.AtomicInteger;


import org.junit._
import Assert._


/**
 * Basic tests for Contrail async API.
 * 
 * @author Ted Stockwell
 */
class ContrailTasksTests {

	@Test def testBasicTask() {
	  
		runTest(new Task() {
			final String expected= "boogity";
			public void execute() throws Pausable, Exception {
				String actual= new ContrailTask<String>() {
					protected String run() throws Pausable, Exception {
						return expected;
					}
				}.submit().get();
				assertEquals(expected, actual);
			}
		});
	}
	

	public void testBasicTaskDomain() throws Throwable {
		runTest(new Task() {
			final private TaskDomain _tracker= new TaskDomain();
			final private TaskDomain.Session _trackerSession= _tracker.beginSession();
			final AtomicInteger count= new AtomicInteger();
			public void execute() throws Pausable, Exception {
				Identifier id= Identifier.create();
				_trackerSession.submit(new ContrailAction(id, Operation.WRITE) {
					protected void action() throws Pausable, Exception {
						ContrailTask.sleep(50);
						count.incrementAndGet();
					}
				});
				_trackerSession.submit(new ContrailAction(id, Operation.FLUSH) {
					protected void action() throws Pausable, Exception {
						assertEquals(1, count.intValue());
						ContrailTask.sleep(50);
						count.incrementAndGet();
					}
				});
				_trackerSession.submit(new ContrailAction(id, Operation.READ) {
					protected void action() throws Pausable, Exception {
						assertEquals(2, count.intValue());
						ContrailTask.sleep(50);
						count.incrementAndGet();
					}
				});
				_trackerSession.complete().get();
				assertEquals(3, count.intValue());
			}
		});
	}

	public void testSleep() throws Throwable {
		runTest(new ContrailAction() {
			@Override
			protected void action() throws Pausable, Exception {
				long start= System.currentTimeMillis();
				ContrailTask.sleep(100);
				assertTrue("ContrailTask.sleep didnt sleep long enough", start+100 <= System.currentTimeMillis());
			}
		});
	}

	/**
	 * At one time there was a bug in the Contrail Async API such that 
	 * if a java.lang.Error was thrown from within a Task, then the associated 
	 * call to IResult.get(), to get the result of the task, did not throw an 
	 * exception.  This test makes sure that throwing a java.lang.Error causes 
	 * an exception to be thrown from the associate Task.get(). 
	 */
	public void testErrorHandling() {
		IResult<Void> result= new ContrailAction() {
			protected void action() throws Pausable, Exception {
				throw new java.lang.Error("catch this");
			}
		}.submit();
		boolean success= true;
		try {
			result.getb();
			success= false;
		}
		catch (Throwable t) {
			// if things are working then we should get here...
		}
		if (!success)
			fail("IResult.getb() should have thrown an exception");
	}
	
}
